core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -383,6 +396,7 @@ public void checkTimeoutNotExhausted()
 
         public void recordRuleInvocation(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos)
         {
+            queryStatsCollector.recordRule(rule, invoked, applied, elapsedNanos);
```

Comment:
it seems that now we will have 3 rule stats collectors. Global, per query for timeout exhaustion and the new per query for query json. Do you think somehow merging those thres together makes sense? We could for example gather global stats through per query collector + use it to control the time.

Reply:
Merge two per query rule statistics.